### DISCUSSION TITLE
Canonical index ordering - treat base tensors special?

### DISCUSSION THREAD

**Discussion by Krzmbrzl** [KoehnLab/lizard#16] [Theory & Implementation]
This is in the context of https://github.com/KoehnLab/lizard/issues/15#issuecomment-1442495508

> we should avoid this by keeping the canonical initial indexing (i.e. the interchange of a and b in the "H" tensor should not happen).

The abba spin case originates as follows:
The original term is
```
H_T1[i⁺j⁺a⁻k⁻](||||) -= H[i⁺j⁺a⁻b⁻](||||) T1[b⁺k⁻](||)
```
which after spin integration has one spin case baba, which reads
```
H_T1[i⁺j⁺a⁻k⁻](\/\/) -= H[i⁺j⁺a⁻b⁻](\/\/) T1[b⁺k⁻](//)
```
transforming `H_T1` into the canonical spin case abba via that i<->j antisymmetry, followed by an index renaming (j->i and i->j) leads to
```
H_T1[i⁺j⁺a⁻k⁻](/\\/) += H[i⁺j⁺b⁻a⁻](/\/\) T1[b⁺k⁻](//)
```
(which coincides with the abba spin case that has already been produced during spin integration of the original term. Therefore, only one of these is kept.)
introducing skeleton tensors here leads to
```
H_T1[i⁺j⁺a⁻k⁻](/\\/) += H[i⁺j⁺b⁻a⁻](....) T1[b⁺k⁻](..)
```

----

As you can see, index canonicalization (in the program from my master's thesis) is performed from left to right, starting with the result tensor. During this, all tensors are treated equally (no one has intrinsic preference over the other). That seemed like a reasonable approach.

@ak-ustutt if I understood you correctly, you would propose to prioritize base tensors like H, T1, T2 and so on when determining the canonical index ordering in a given term, correct? So instead of going left to right, we would process tensors by their priority and only after that from left to right.

And in addition to this, we would also require prioritizing index names over spin case when determining the canonical order. However, iirc I specifically prioritized the spin case over index names, because otherwise the spin case was rarely changed during canonicalization and then later in the code (spin summation) we would require everything to be in specific spin cases, in order for skeleton tensors to be introduced. Thus, it seemed simpler to make sure we arrive at the desired spin cases as soon (and as consistently) as possible, without having to reinvent the canonicalization routine for spin summation :thinking:

**Comment by ak-ustutt** [comment]
As the "canonical" spin case we should better define ```/\/\``` as this corresponds to the skeleton quantities for the 4-index case

**Reply by Krzmbrzl** [reply]
That is what is aimed for, but H_T1 does not have the required symmetries to be transformed to this case (only i and j can be exchanged as a and k come from different tensors).

**Reply by ak-ustutt** [reply]
```
H_T1[i⁺j⁺a⁻k⁻](/\\/) += H[i⁺j⁺b⁻a⁻](/\/\) T1[b⁺k⁻](//)
```
can be written as:
```
H_T1[j⁺i⁺a⁻k⁻](\/\/) -= H[i⁺j⁺b⁻a⁻](/\/\) T1[b⁺k⁻](//)
```
or after a global spin flip
```
H_T1[j⁺i⁺a⁻k⁻](/\/\) -= H[i⁺j⁺b⁻a⁻](\/\/) T1[b⁺k⁻](\\)
```
which is equivalent to the previous in terms of skeleton quantities.
So this should be possible in most cases, but there clearly can be more complex ones, like INT[ecec] ... Here, the exchange terms live on a tensor INTX[ecce]. I think that we should accept that we have to consider both INT and INTX (as this is closest to how we treat these later in ITF).
Your observation on indices coming from different tensors is important (although not to relevant in this specific case). It plays a role, when we have e.g. INT[e1 e2 cc] where e1 and e2 are both "external" but come from different tensors and if e1 and e2 get afterwards contracted to non-equivalent indices (does not really happen for the CC amplitude equations but for the Lambda equations).
